### PR TITLE
Add redis state provider

### PR DESCRIPTION
**Pull request by RemindD**
Use Redis HSet to restore key-value pairs. Have Upsert, Get, List, Delete implemented.

- key: namespace:resource.group:objName
- fields: etag and values

### CODE REVIEW

File: coa/pkg/apis/v1alpha2/providers/states/redisstate/redisstate.go [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+			if err != nil || len(result) == 0 {
+				rLog.Errorf("  P (Redis State): failed to get entry for key %s: %+v", key, err)
+				continue
+			}
+			parts := strings.Split(key, separator)
+			if len(parts) != 3 {
+				rLog.Errorf("  P (Redis State): key is not valid %s: %+v", key, err)
+				continue
+			}
+			entry, err := CastRedisPropertiesToStateEntry(parts[2], result)
```
Comment (Haishi2016) [inline coa/pkg/apis/v1alpha2/providers/states/redisstate/redisstate.go:182]: We need to support filters in list quest. Please see memory state provider or k8s state provider as an example

**Review by Haishi2016 — DISMISSED** [review]

File: coa/pkg/apis/v1alpha2/providers/states/redisstate/redisstate.go (hunk 1 above)
Comment (RemindD) [reply coa/pkg/apis/v1alpha2/providers/states/redisstate/redisstate.go:182]: Thanks Haishi for bringing this up. I have looked into secondary index for Redis previously and it's supported by RediSearch which is published by Redis Lab and uses Redis Source Available license. So it's not proper to use the product in Symphony. If we want to have the filters, I am afraid we need to list all the items into memory and do filter by symphony. Do you think it's acceptable or not?

Comment (Haishi2016) [reply coa/pkg/apis/v1alpha2/providers/states/redisstate/redisstate.go:182]: Behavior consistency is important as other parts of the system would expect filters to work when Redis is used. We can worry about performance at a later time.

Comment (RemindD) [reply coa/pkg/apis/v1alpha2/providers/states/redisstate/redisstate.go:224]: Add filtering in List interface using shared in-memory filter as memory state provider. Also respect UpdateStateOnly field in Upsert interface.

File: api/pkg/apis/v1alpha1/providers/states/k8s/k8s.go [hunk 2]
```diff
@@ -253,7 +253,7 @@ func (s *K8sStateProvider) Upsert(ctx context.Context, entry states.UpsertReques
 			item.SetAnnotations(metadata.Annotations)
 		}
 		getResourceVersion := false
-		if v, ok := dict["spec"]; ok {
+		if v, ok := dict["spec"]; ok && !entry.Options.UpdateStateOnly {
```
Comment (msftcoderdjw) [inline api/pkg/apis/v1alpha1/providers/states/k8s/k8s.go:256]: Can you please hint me why we need the new option - UpdateStateOnly?

Comment (RemindD) [reply api/pkg/apis/v1alpha1/providers/states/k8s/k8s.go:256]: UpsertOption has a field UpdateStateOnly but k8s state provider didn't respect it previously. We should only update status only if UpdateStateOnly is true. You can refer to memoryStateProvider for more information.

File: coa/pkg/apis/v1alpha2/providers/states/redisstate/redisstate.go [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+			if err != nil {
+				rLog.Errorf("  P (Redis State): failed to cast entry for key %s: %+v", key, err)
+				continue
+			}
+			if request.FilterType != "" && request.FilterValue != "" {
+				var match bool
+				match, err = states.MatchFilter(entry, request.FilterType, request.FilterValue)
```
Comment (msftcoderdjw) [inline coa/pkg/apis/v1alpha2/providers/states/redisstate/redisstate.go:231]: I can understand your concern that RediSearch is under (i) Redis Source Available License 2.0 (RSALv2) or (ii) the Server Side Public License v1 (SSPLv1), not all Symphony deployment can leverage RediSearch. Under this case, we can only filter data property in client side. But there's still some follow up items we need to consider in future,
1) Using client side filter, how can we support paging if needed in future?
2) If Symphony is deployed with standalone and users want to supply a redis server with RediSearch loaded, we can extend redis state provider to leverage this feature.

Considering the need of above two follow up tasks are not in vision, I am OK we can firstly handle the redis state filter in memory to unblock the crash consistency issue which has higher priority.

**Review by msftcoderdjw — COMMENTED** [review]
> UpsertOption has a field UpdateStateOnly but k8s state provider didn't respect it previously. We should only update status only if UpdateStateOnly is true. You can refer to memoryStateProvider for more information. 

Thanks for calling this out. Seems we should change the naming as UpdateStatusOnly for easier understanding.

File: api/pkg/apis/v1alpha1/providers/providerfactory_test.go [hunk 4]
```diff
@@ -284,6 +284,11 @@ func TestCreateProviderForTargetRole(t *testing.T) {
 							Provider: "providers.state.k8s",
 							Config:   map[string]string{},
 						},
+						{
```
[the remark this answers is not in the document]
Comment (RemindD) [reply api/pkg/apis/v1alpha1/providers/providerfactory_test.go:297]: Added test for TestProvider function. Removed redis state provider in target provider as the provider is not expected for target.

File: coa/pkg/apis/v1alpha2/providers/states/redisstate/redisstate.go (hunk 3 above)
Comment (RemindD) [reply coa/pkg/apis/v1alpha2/providers/states/redisstate/redisstate.go:231]: Issue has been opened per offline discussion.

File: coa/pkg/apis/v1alpha2/providers/states/redisstate/redisstate.go [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+	options := &redis.Options{
+		Addr:            r.Config.Host,
+		Password:        r.Config.Password,
+		DB:              0,
+		MaxRetries:      3,
+		MaxRetryBackoff: time.Second * 2,
+	}
+	if r.Config.RequiresTLS {
+		options.TLSConfig = &tls.Config{
+			InsecureSkipVerify: !r.Config.RequiresTLS,
```
Comment (Haishi2016) [inline coa/pkg/apis/v1alpha2/providers/states/redisstate/redisstate.go:111]: I think require TLS and skipping certificate validation are two separate things - one to indicate if TLS is needed, the other indicates if certificate validation is skipped (such as when an untrusted self-issued certificate is used)?

Comment (RemindD) [reply coa/pkg/apis/v1alpha2/providers/states/redisstate/redisstate.go:111]: I copied it from Redis pubsub configuration. Maybe we can change the configuration in both places when fixing issue https://github.com/eclipse-symphony/symphony/issues/253?

**Review by Haishi2016 — APPROVED** [review]

File: api/pkg/apis/v1alpha1/providers/providerfactory_test.go [hunk 6]
```diff
@@ -72,6 +74,14 @@ func TestCreateProvider(t *testing.T) {
 		assert.NotNil(t, *provider.(*k8sstate.K8sStateProvider))
 	}
 
+	if testRedis == "" {
+		t.Log("Skipping providers.state.redis test as TEST_MINIKUBE_ENABLED is not set")
```
[the remark this answers is not in the document]
Comment (RemindD) [reply api/pkg/apis/v1alpha1/providers/providerfactory_test.go]: fixed

**Review by msftcoderdjw — APPROVED** [review]

**Pull request merged**